src/main/scala/org/allenai/pipeline/FileItem.scala

```diff
@@ -0,0 +1,70 @@
+package org.allenai.pipeline
+
+import com.amazonaws.services.s3.{ AmazonS3Client, AmazonS3 }
+
+import scala.reflect.ClassTag
+
+import java.io.File
+
+/** File-like object that can be checked for existence / read / written.
+  * Abstracts over an absolute path and a serialization method.
+  * Examples: flat file (single file), RDD (multiple files).
+  *
+  * The content of a FileItem is unreliable in presence of write interruptions.
+  * Use a Producer[T].persisted(...) to ensure write atomicity.
+  */
+trait FileItem[T] {
```

Comment:
I could live with Artifact as well. It's a standard notion for build systems, and this is a build system for data disguised as a Scala library ;)

===

On separate read and write. I'll be the first to point out that clients should take objects with unused methods as arguments. However, the tradeoff is that it becomes a mouthful to separate the exists/read/write views. And in practice, there will never be an PersistedItem that doesn't offer all three views, otherwise it's useless.
 
I tend to see the PersistedItem as a refined chunk of the Storage (formerly known as FileSystem). A hypothetical PersistedItemReader would be obtained as persistedItem.reader() and used as persistedItemReader.read(), but at that point we might as well write persistedItem.read() directly.

Best to experiment with different API variants on concrete use-cases.

PS. Java File API is a complete mess. Think of each new FileInputStream as a big fat dependencies on global objects that's hard to fake out in tests.